### DISCUSSION TITLE
Lots of WebGL / Shader related errors

### DISCUSSION THREAD

**Discussion by dottodot** [visgl/deck.gl#5202] [Q&A]
I'm noticing a lot of errors on our site relating to WebGL and Shader issues and just trying to understand what's causing them. Is it down to WebGL not being supported / unsuitable GPU specs, and does deck.gl gracefully fullback to an unanimated version? Here's an example of some of the errors.

WebGL context: Failed to create a WebGL2 context.
GLSL compilation errors in fragment shader solid-polygon-layer-fragment-shader-9
WebGL context: Could not create a WebGL context, VENDOR = 0xffff, DEVICE = 0xffff, Sandboxed = no, Optimus = no, AMD switchable = no, Reset notification strategy = 0x0000, ErrorMessage = OffscreenContext Creation failed, GpuChannelHost creation failed
WebGL context: WebGL creation failed: * AllowWebgl2:false restricts context creation on this system.
WebGL instanced rendering not supported

**Comment by Pessimistress** [comment]
Do you know the browser/OS from which these errors are generated? If the browser disables WebGL then there is nothing we can fallback to. You may specify the `onError` prop and catch these in your app.

**Comment by dottodot** [comment]
It's generally quite a variety of browsers. Amazon Silk seems to be the most common offender, however I'm also seeing it for version 86,87 of Chrome desktop and android which I find surprising

**Comment by Pessimistress** [comment]
I see an old issue thread reporting Amazon Silk not able to use WebGL: https://forums.developer.amazon.com/questions/178642/webview-web-app-runtime-webgl-support-for-kindle-7.html

The general strategy of deck.gl is to attempt creating a WebGL2 context if the browser supports it; if it fails, fallback to WebGL1. There is nothing we can do if WebGL is not supported at all, which could be due to an outdated browser, hardware not compatible with the browser, or a restricted runtime environment that disables WebGL via flags. I suggest you use `onError` to capture these scenarios and display a warning/alternative view when it happens.

**Comment by dottodot** [comment]
Should this work? It doesn't seem to be catch the error. Tried in firefox with webgl disabled and the page is throwing error just not through onError
```
 <DeckGL onError={(error) => console.log('catch error',  error)}>
```

**Comment by Pessimistress** [comment]
Looks like `onError` is not properly catching the WebGL context creation exceptions. I'll put up a fix.

**Comment by Pessimistress** [comment]
This is fixed in `deck.gl@8.3.11` and `luma.gl@8.3.2` (you may need to remove the lock file to upgrade the luma.gl dependency).

**Comment by dottodot** [comment]
@Pessimistress Thanks for the update, the errors are now being caught in `onError`

**Comment by Jamal8548** [comment]
I have had a similar issue with an Nvidia Quadro FX1800 but have found a workaround.

Go to chrome://flags and look for Choose ANGLE graphics backend option. I tried out each option to see if the fault occurred on every setting. It didn't thankfully, but some where much slower to render than others.

I would check the options available to you here and try out each of them. Hopefully at least one should work for you.
